### PR TITLE
Freepay cancel or refund

### PR DESCRIPTION
**Pull request by rus-daniel**

### CODE REVIEW

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 1]
```diff
@@ -9,11 +9,19 @@
  * @copyright Copyright (c) 2017 Oyst (http://www.oyst.com)
  */
 
+use Oyst\Api\OystApiClientFactory;
+use Oyst\Api\OystPaymentApi;
+use Oyst\Classes\OystPrice;
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: add break line after last use;

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 2]
```diff
@@ -24,8 +32,27 @@ class Oyst_OneClick_Model_Payment_ApiWrapper extends Mage_Core_Model_Abstract
     public function getPaymentUrl($params)
     {
         /** @var Oyst_OneClick_Model_Api $paymentApi */
-        $paymentApi = Mage::getModel('oyst_oneclick/api');
-        $response = $paymentApi->sendPayment(Oyst_OneClick_Model_Api::TYPE_PAYMENT, $params)->getResponse();
+        $api = Mage::getModel('oyst_oneclick/api');
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Move this in a constructor with the same idea of ```Model/OneClick/ApiWrapper.php```

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 3]
```diff
@@ -37,11 +64,33 @@ public function getPaymentUrl($params)
      */
     public function cancelOrRefund($lastTransId, $amount = null)
     {
-        /** @var Oyst_OneClick_Model_Api $paymentApi */
-        $paymentApi = Mage::getModel('oyst_oneclick/api');
+        $type = Oyst_OneClick_Model_Api::TYPE_PAYMENT;
+
+        /** @var Oyst_OneClick_Model_Api $api */
+        $api = Mage::getModel('oyst_oneclick/api');
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Move this in a constructor with the same idea of ```Model/OneClick/ApiWrapper.php```

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 4]
```diff
@@ -24,8 +32,27 @@ class Oyst_OneClick_Model_Payment_ApiWrapper extends Mage_Core_Model_Abstract
     public function getPaymentUrl($params)
     {
         /** @var Oyst_OneClick_Model_Api $paymentApi */
-        $paymentApi = Mage::getModel('oyst_oneclick/api');
-        $response = $paymentApi->sendPayment(Oyst_OneClick_Model_Api::TYPE_PAYMENT, $params)->getResponse();
+        $api = Mage::getModel('oyst_oneclick/api');
+
+        $type = Oyst_OneClick_Model_Api::TYPE_PAYMENT;
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Use const for this.

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 5]
```diff
@@ -37,11 +64,33 @@ public function getPaymentUrl($params)
      */
     public function cancelOrRefund($lastTransId, $amount = null)
     {
-        /** @var Oyst_OneClick_Model_Api $paymentApi */
-        $paymentApi = Mage::getModel('oyst_oneclick/api');
+        $type = Oyst_OneClick_Model_Api::TYPE_PAYMENT;
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Use const for this.

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 6]
```diff
@@ -24,8 +32,27 @@ class Oyst_OneClick_Model_Payment_ApiWrapper extends Mage_Core_Model_Abstract
...
-        $paymentApi = Mage::getModel('oyst_oneclick/api');
-        $response = $paymentApi->sendPayment(Oyst_OneClick_Model_Api::TYPE_PAYMENT, $params)->getResponse();
+        $api = Mage::getModel('oyst_oneclick/api');
+
+        $type = Oyst_OneClick_Model_Api::TYPE_PAYMENT;
+
+        $apiKey = Mage::getStoreConfig('payment/oyst_abstract/api_login');
+        $userAgent = $api->getUserAgent();
+        $env = Mage::getStoreConfig('payment/oyst_abstract/mode');
+        $url = $api->getCustomApiUrl('payment/oyst_abstract/');
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: All previous block, is not needed if you use ```getClient``` of Api.php

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 7]
```diff
@@ -24,8 +32,27 @@ class Oyst_OneClick_Model_Payment_ApiWrapper extends Mage_Core_Model_Abstract
...
+
+        /** @var OystPaymentAPI $oystClient */
+        $oystClient = OystApiClientFactory::getClient($type, $apiKey, $userAgent, $env, $url);
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Same as above.

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 8]
```diff
@@ -37,11 +64,33 @@ public function getPaymentUrl($params)
...
-        $paymentApi = Mage::getModel('oyst_oneclick/api');
+        $type = Oyst_OneClick_Model_Api::TYPE_PAYMENT;
+
+        /** @var Oyst_OneClick_Model_Api $api */
+        $api = Mage::getModel('oyst_oneclick/api');
+
+        $apiKey = Mage::getStoreConfig('payment/oyst_abstract/api_login');
+        $userAgent = $api->getUserAgent();
+        $env = Mage::getStoreConfig('payment/oyst_abstract/mode');
+        $url = $api->getCustomApiUrl('payment/oyst_abstract/');
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Useless block, use ```getClient``` in constructor like in ```Model/OneClick/ApiWrapper.php```

File: app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php [hunk 9]
```diff
@@ -37,11 +64,33 @@ public function getPaymentUrl($params)
...
+            $oystClient = OystApiClientFactory::getClient($type, $apiKey, $userAgent, $env);
+        } else {
+            $oystClient = OystApiClientFactory::getClient($type, $apiKey, $userAgent);
+        }
+
+        if (!(null === $amount)) {
+            $amount = new OystPrice($amount, self::CURRENCY);
+        }
+
+        $oystClient->cancelOrRefund($lastTransId, $amount);
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Payment/ApiWrapper.php]: Recursive call !!! Why ?

File: app/code/community/Oyst/OneClick/etc/config.xml [hunk 10]
```diff
@@ -113,37 +102,6 @@
         </layout>
     </frontend>
     <adminhtml>
-        <events>
-            <!-- Show/Hide the Button 1-Click Tab -->
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/config.xml]: Why this block has been removed?

File: app/code/community/Oyst/OneClick/Model/Adminhtml/Observer.php [hunk 11]
```diff
@@ -14,39 +14,6 @@
  */
 class Oyst_OneClick_Model_Adminhtml_Observer extends Mage_Core_Model_Abstract
 {
-    /**
-     * Custom cancel and refund
-     *
-     * @return
-     */
-    public function cancelAndRefundButton(Varien_Event_Observer $observer)
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Adminhtml/Observer.php]: As I remember this function is for OC, so is shouldn't be changes in this PR?

**Comment by rbenezra** [comment]
OK QA

**Pull request merged**